LibWeb: Allow reordering of keywords in CSS positions

shannonbooth opened the pull request:
As the spec points out:
> Note that a pair of keywords can be reordered while a combination of
> keyword and length or percentage cannot. So center left is valid while
> 50% left is not.

This was a bug in our implementation of alternative 2 of css-values-3,
resulting in the following CSS failing to be parsed:

`background-position: center right;`

This fixes the issue as part of an update of the parsing to
css-values-4. As far as I can tell, the grammar is equivalent - but
simpler to implement due to the lack of optional values.

The fix for this issue is also as part of alternative 2 parsing in the
new grammar.

Progress towards: #22401

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -2629,140 +2629,176 @@ RefPtr<PositionStyleValue> Parser::parse_position_value(TokenStream<ComponentVal
+        };
+
+        // [ left | center | right | <length-percentage> ]
+        RefPtr<EdgeStyleValue> horitontal_edge = parse_position_or_length(true);
```
Typo: Horizontal not horitontal.

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -2629,140 +2629,176 @@ RefPtr<PositionStyleValue> Parser::parse_position_value(TokenStream<ComponentVal
         auto transaction = tokens.begin_transaction();
+
+        tokens.skip_whitespace();
+        auto token = tokens.next_token();
```
Store tokens in an `auto&` to avoid copying them.

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -2629,140 +2629,176 @@ RefPtr<PositionStyleValue> Parser::parse_position_value(TokenStream<ComponentVal
-        }
+        auto parse_position_or_length = [&](bool as_horizontal) -> RefPtr<EdgeStyleValue> {
+            tokens.skip_whitespace();
+            auto token = tokens.next_token();
```
`auto&`

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -2629,140 +2629,176 @@ RefPtr<PositionStyleValue> Parser::parse_position_value(TokenStream<ComponentVal
-            auto edge = parse_position_edge(tokens.next_token());
-            if (!edge.has_value() || !is_horizontal(*edge, false))
-                return false;
+            auto first_token = tokens.next_token();
+
+            auto maybe_position = parse_position_edge(first_token);
```
Here we can avoid the `first_token` variable altogether and just pass `tokens.next_token()` to `parse_position_edge()`.

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -2629,140 +2629,176 @@ RefPtr<PositionStyleValue> Parser::parse_position_value(TokenStream<ComponentVal
     // [ [ top | bottom ] <length-percentage> ]
-    auto alternative_3 = [&]() -> RefPtr<PositionStyleValue> {
-        auto transaction = tokens.begin_transaction();
-        tokens.skip_whitespace();
-        RefPtr<EdgeStyleValue> horizontal_edge;
-        RefPtr<EdgeStyleValue> vertical_edge;
+    auto alternative_4 = [&]() -> RefPtr<PositionStyleValue> {
+        struct PositionAndLength {
+            PositionEdge position;
+            LengthPercentage length;
+        };
 
-        auto parse_horizontal = [&] {
-            // [ left | right ] <length-percentage> ]
-            auto transaction = tokens.begin_transaction();
+        auto parse_position_and_length = [&]() -> Optional<PositionAndLength> {
             tokens.skip_whitespace();
-            auto edge = parse_position_edge(tokens.next_token());
-            if (!edge.has_value() || !is_horizontal(*edge, false))
-                return false;
+            auto first_token = tokens.next_token();
+
+            auto maybe_position = parse_position_edge(first_token);
+            if (!maybe_position.has_value())
+                return {};
 
             tokens.skip_whitespace();
-            auto length_percentage = parse_length_percentage(tokens.next_token());
-            if (!length_percentage.has_value())
-                return false;
+            auto second_token = tokens.next_token();
 
-            horizontal_edge = EdgeStyleValue::create(*edge, *length_percentage);
-            transaction.commit();
-            return true;
+            auto maybe_length = parse_length_percentage(second_token);
```
Same here.

AtkinsSJ reviewed (changes requested):
This looks great! Just a few nit-picks.

shannonbooth commented:
Thanks for the review! Addressed all comments, and also finally managed to convince my window to be a bit narrower for the updated ref-test image (for some reason it doesn't like going that narrow)
